Linear Regression - Data Cleanup
================================

Here we get into the detailed steps involved in data cleanup of the Lending Club dataset. We glossed over this in the Data Exploration lesson and left it as an exercise.  This is where we actually do the cleanup step by step.


1. Remove '%'
 * strip the last letter
 * map to float


2. Remove ' months' suffix
 * strip the tailing letters
 * map to int

3. Remove bad data
 * n/a values
 * outlier income
  * histograms
  * describe

4. Map FICO.Range to single int
 * split on '-'
 * use bottom value
 * convert to int
 * show how to do mid-point



In [1]:
%pylab inline
import pandas as pd
from numpy import nan as NA
loansData = pd.read_csv('https://spark-public.s3.amazonaws.com/dataanalysis/loansData.csv')
loansData

Populating the interactive namespace from numpy and matplotlib


Amount.Requested  Amount.Funded.By.Investors Interest.Rate  \
81174              20000                    20000.00         8.90%   
99592              19200                    19200.00        12.12%   
80059              35000                    35000.00        21.98%   
15825              10000                     9975.00         9.99%   
33182              12000                    12000.00        11.71%   
62403               6000                     6000.00        15.31%   
48808              10000                    10000.00         7.90%   
22090              33500                    33450.00        17.14%   
76404              14675                    14675.00        14.33%   
15867               7000                     7000.00         6.91%   
94971               2000                     2000.00        19.72%   
36911              10625                    10625.00        14.27%   
41200              28000                    27975.00        21.67%   
83869              35000                    34950.00         8.90%   
53853               9600                     9600.00         7.62%   
21399              25000                    24975.00        15.65%   
62127              10000                    10000.00        12.12%   
23446              14000                    13900.25        10.37%   
44987              10000                    10000.00         9.76%   
17977               5200                     5175.00         9.99%   
86099              22000                    21975.00        21.98%   
99483              30000                    30000.00        19.05%   
28798               6500                     6500.00        17.99%   
24168              17400                    17400.00        11.99%   
10356               4000                     4000.00        16.82%   
46027               7200                     7200.00         7.90%   
2238                8000                     8000.00        14.42%   
65278               8000                     8000.00        15.31%   
4227                3000                     3000.00         8.59%   
50182              14500                    14500.00         7.90%   
...                  ...                         ...           ...   
84265              20000                    20000.00        22.95%   
80231              19000                    19000.00         7.90%   
49533              17300                    17250.00        22.45%   
102514              7000                      711.54        15.13%   
78618               7200                     7200.00        18.75%   
86953              10000                    10000.00        14.09%   
80129               4000                     3925.00        14.09%   
85216              17500                    17500.00         8.90%   
38247              20000                    20000.00        11.71%   
91245              16200                    16200.00        15.80%   
53041              10000                    10000.00         6.03%   
63051              27000                    27000.00         6.62%   
14446               4500                     4475.00         7.51%   
68628              15875                    15875.00        14.33%   
98758              15000                    15000.00        10.16%   
13070              25000                    24950.00        10.75%   
45836               7000                     7000.00        17.27%   
52330              15000                    15000.00        19.99%   
48243              17000                    17000.00        15.81%   
63256              19075                    19075.00        18.75%   
42124              10000                    10000.00        11.71%   
78043               8475                     8475.00         7.62%   
925                 6400                     6350.00        10.08%   
74047              30000                    30000.00        23.28%   
49957              24000                    23975.00        14.65%   
23735             

In [2]:
loansData['Interest.Rate'][0:5] # first five rows of Interest.Rate

81174     8.90%
99592    12.12%
80059    21.98%
15825     9.99%
33182    11.71%
Name: Interest.Rate, dtype: object

* removing % signs from rates

In [3]:
ir = loansData['Interest.Rate']
ir[0:5]

81174     8.90%
99592    12.12%
80059    21.98%
15825     9.99%
33182    11.71%
Name: Interest.Rate, dtype: object

In [4]:
type(ir[0:1])
irbak = ir
type(irbak)
loansDataBak = loansData
loansDataBak.reset_index()
loansDataBak
loansDataBak['Interest.Rate']
loansData

Amount.Requested  Amount.Funded.By.Investors Interest.Rate  \
81174              20000                    20000.00         8.90%   
99592              19200                    19200.00        12.12%   
80059              35000                    35000.00        21.98%   
15825              10000                     9975.00         9.99%   
33182              12000                    12000.00        11.71%   
62403               6000                     6000.00        15.31%   
48808              10000                    10000.00         7.90%   
22090              33500                    33450.00        17.14%   
76404              14675                    14675.00        14.33%   
15867               7000                     7000.00         6.91%   
94971               2000                     2000.00        19.72%   
36911              10625                    10625.00        14.27%   
41200              28000                    27975.00        21.67%   
83869              35000                    34950.00         8.90%   
53853               9600                     9600.00         7.62%   
21399              25000                    24975.00        15.65%   
62127              10000                    10000.00        12.12%   
23446              14000                    13900.25        10.37%   
44987              10000                    10000.00         9.76%   
17977               5200                     5175.00         9.99%   
86099              22000                    21975.00        21.98%   
99483              30000                    30000.00        19.05%   
28798               6500                     6500.00        17.99%   
24168              17400                    17400.00        11.99%   
10356               4000                     4000.00        16.82%   
46027               7200                     7200.00         7.90%   
2238                8000                     8000.00        14.42%   
65278               8000                     8000.00        15.31%   
4227                3000                     3000.00         8.59%   
50182              14500                    14500.00         7.90%   
...                  ...                         ...           ...   
84265              20000                    20000.00        22.95%   
80231              19000                    19000.00         7.90%   
49533              17300                    17250.00        22.45%   
102514              7000                      711.54        15.13%   
78618               7200                     7200.00        18.75%   
86953              10000                    10000.00        14.09%   
80129               4000                     3925.00        14.09%   
85216              17500                    17500.00         8.90%   
38247              20000                    20000.00        11.71%   
91245              16200                    16200.00        15.80%   
53041              10000                    10000.00         6.03%   
63051              27000                    27000.00         6.62%   
14446               4500                     4475.00         7.51%   
68628              15875                    15875.00        14.33%   
98758              15000                    15000.00        10.16%   
13070              25000                    24950.00        10.75%   
45836               7000                     7000.00        17.27%   
52330              15000                    15000.00        19.99%   
48243              17000                    17000.00        15.81%   
63256              19075                    19075.00        18.75%   
42124              10000                    10000.00        11.71%   
78043               8475                     8475.00         7.62%   
925                 6400                     6350.00        10.08%   
74047              30000                    30000.00        23.28%   
49957              24000                    23975.00        14.65%   
23735             

In [5]:
ldb = loansDataBak.reset_index() # explain
ldb['Interest.Rate'][0:5]
irates = ldb['Interest.Rate'][0:]
type(irates)

[irates[k] for k in (0,1,2,3,4)]
#int_rates = [ float(x[:-1]) for x in irates[k] ]
#"""
#---------------------------------------------------------------------------
#ValueError                                Traceback (most recent call last)
#<ipython-input-46-5c4f2333efd1> in <module>()
#      5 
#      6 [irates[k] for k in (0,1,2,3,4)]
#----> 7 int_rates = [ float(x[:-1]) for x in irates[k] ]
#
#ValueError: could not convert string to float: 
#"""



['8.90%', '12.12%', '21.98%', '9.99%', '11.71%']

In [6]:
type(ldb)

pandas.core.frame.DataFrame

In [7]:
srates = ldb['Interest.Rate']

In [8]:
#nas = [ x for x in srates if x.isnull() ] # AttributeError: 'str' object has no attribute 'isnull'

In [9]:
nas = [ x for x in srates if not(x[0].isdigit()) ] # AttributeError: 'str' object has no attribute 'isnull'

In [10]:
len(nas)

0

In [11]:
srates[0][:-1]

'8.90'

In [12]:
float(srates[0][:-1])

8.9

In [13]:
nopct = [ x[:-1] for x in srates ]

In [14]:
flrates = [float(x) for x in nopct]

In [15]:
flrates[0:5]

[8.9, 12.12, 21.98, 9.99, 11.71]

In [21]:
ldb

index  Amount.Requested  Amount.Funded.By.Investors Interest.Rate  \
0      81174             20000                    20000.00         8.90%   
1      99592             19200                    19200.00        12.12%   
2      80059             35000                    35000.00        21.98%   
3      15825             10000                     9975.00         9.99%   
4      33182             12000                    12000.00        11.71%   
5      62403              6000                     6000.00        15.31%   
6      48808             10000                    10000.00         7.90%   
7      22090             33500                    33450.00        17.14%   
8      76404             14675                    14675.00        14.33%   
9      15867              7000                     7000.00         6.91%   
10     94971              2000                     2000.00        19.72%   
11     36911             10625                    10625.00        14.27%   
12     41200             28000                    27975.00        21.67%   
13     83869             35000                    34950.00         8.90%   
14     53853              9600                     9600.00         7.62%   
15     21399             25000                    24975.00        15.65%   
16     62127             10000                    10000.00        12.12%   
17     23446             14000                    13900.25        10.37%   
18     44987             10000                    10000.00         9.76%   
19     17977              5200                     5175.00         9.99%   
20     86099             22000                    21975.00        21.98%   
21     99483             30000                    30000.00        19.05%   
22     28798              6500                     6500.00        17.99%   
23     24168             17400                    17400.00        11.99%   
24     10356              4000                     4000.00        16.82%   
25     46027              7200                     7200.00         7.90%   
26      2238              8000                     8000.00        14.42%   
27     65278              8000                     8000.00        15.31%   
28      4227              3000                     3000.00         8.59%   
29     50182             14500                    14500.00         7.90%   
...      ...               ...                         ...           ...   
2470   84265             20000                    20000.00        22.95%   
2471   80231             19000                    19000.00         7.90%   
2472   49533             17300                    17250.00        22.45%   
2473  102514              7000                      711.54        15.13%   
2474   78618              7200                     7200.00        18.75%   
2475   86953             10000                    10000.00        14.09%   
2476   80129              4000                     3925.00        14.09%   
2477   85216             17500                    17500.00         8.90%   
2478   38247             20000                    20000.00        11.71%   
2479   91245             16200                    16200.00        15.80%   
2480   53041             10000                    10000.00         6.03%   
2481   63051             27000                    27000.00         6.62%   
2482   14446              4500                     4475.00         7.51%   
2483   68628             15875                    15875.00        14.33%   
2484   98758             15000                    15000.00        10.16%   
2485   13070             25000                    24950.00        10.75%   
2486   45836              7000                     7000.00        17.27%   
2487   52330             15000                    15000.00        19.99%   
2488   48243             17000                    17000.00        15.81%   
2489   63256             19075                    19075.00        18.75%   
2490   42124             10000                    10000

In [25]:
ldb['Interest.Rate'] = flrates

In [26]:
ldb[0:5]

index  Amount.Requested  Amount.Funded.By.Investors  Interest.Rate  \
0  81174             20000                     20000.0           8.90   
1  99592             19200                     19200.0          12.12   
2  80059             35000                     35000.0          21.98   
3  15825             10000                      9975.0           9.99   
4  33182             12000                     12000.0          11.71   

  Loan.Length        Loan.Purpose Debt.To.Income.Ratio State Home.Ownership  \
0   36 months  debt_consolidation               14.90%    SC       MORTGAGE   
1   36 months  debt_consolidation               28.36%    TX       MORTGAGE   
2   60 months  debt_consolidation               23.81%    CA       MORTGAGE   
3   36 months  debt_consolidation               14.30%    KS       MORTGAGE   
4   36 months         credit_card               18.78%    NJ           RENT   

   Monthly.Income FICO.Range  Open.CREDIT.Lines  Revolving.CREDIT.Balance  \
0         6541.67    735-739               14.0                   14272.0   
1         4583.33    715-719               12.0                   11140.0   
2        11500.00    690-694               14.0                   21977.0   
3         3833.33    695-699               10.0                    9346.0   
4         3195.00    695-699               11.0                   14469.0   

   Inquiries.in.the.Last.6.Months Employment.Length  
0                             2.0          < 1 year  
1                             1.0           2 years  
2                             1.0           2 years  
3                             0.0           5 years  
4                             0.0           9 years

In [27]:
ldb['Interest.Rate'][0:5]

0     8.90
1    12.12
2    21.98
3     9.99
4    11.71
Name: Interest.Rate, dtype: float64

In [28]:
srates = loansData['Interest.Rate']

In [29]:
nopct = [ x[:-1] for x in srates ]

In [30]:
flrates = [float(x) for x in nopct]


In [31]:
rates = [float(x[:-1]) for x in srates] # use this

In [32]:
flrates == rates

True

In [33]:
loansData['Interest.Rate'] = flrates

In [34]:
loansData['Interest.Rate'][0:5]

81174     8.90
99592    12.12
80059    21.98
15825     9.99
33182    11.71
Name: Interest.Rate, dtype: float64

### Conclusion of step 1
* ok! whew! we're done with the % symbol stuff
* we learnt quite a few things along the way that will be useful in the next part

## Step 2: Remove the months

In [35]:
withmons = ldb['Loan.Length'] 

In [36]:
wmons = withmons[0:]

In [37]:
wmons[0:5]

0    36 months
1    36 months
2    60 months
3    36 months
4    36 months
Name: Loan.Length, dtype: object

In [38]:
wmons

0       36 months
1       36 months
2       60 months
3       36 months
4       36 months
5       36 months
6       36 months
7       60 months
8       36 months
9       36 months
10      36 months
11      36 months
12      60 months
13      36 months
14      36 months
15      60 months
16      36 months
17      60 months
18      36 months
19      60 months
20      36 months
21      60 months
22      60 months
23      36 months
24      60 months
25      36 months
26      36 months
27      36 months
28      36 months
29      36 months
          ...    
2470    60 months
2471    36 months
2472    60 months
2473    36 months
2474    36 months
2475    36 months
2476    36 months
2477    36 months
2478    36 months
2479    60 months
2480    36 months
2481    36 months
2482    36 months
2483    36 months
2484    36 months
2485    36 months
2486    36 months
2487    36 months
2488    36 months
2489    36 months
2490    36 months
2491    36 months
2492    36 months
2493    60 months
2494    36

In [39]:
wmons[0].split()

['36', 'months']

In [40]:
wmons[0].split()[0]

'36'

In [41]:
int(wmons[0].split()[0])

36

In [42]:
x = wmons[0].split()


In [43]:
x[0]

'36'

In [44]:
int(x[0])

36

In [45]:
intmons = [ int(x.split()[0]) for x in wmons ] 

In [46]:
intmons[0:10]

[36, 36, 60, 36, 36, 36, 36, 60, 36, 36]

In [47]:
loansData['Loan.Length']

81174     36 months
99592     36 months
80059     60 months
15825     36 months
33182     36 months
62403     36 months
48808     36 months
22090     60 months
76404     36 months
15867     36 months
94971     36 months
36911     36 months
41200     60 months
83869     36 months
53853     36 months
21399     60 months
62127     36 months
23446     60 months
44987     36 months
17977     60 months
86099     36 months
99483     60 months
28798     60 months
24168     36 months
10356     60 months
46027     36 months
2238      36 months
65278     36 months
4227      36 months
50182     36 months
            ...    
84265     60 months
80231     36 months
49533     60 months
102514    36 months
78618     36 months
86953     36 months
80129     36 months
85216     36 months
38247     36 months
91245     60 months
53041     36 months
63051     36 months
14446     36 months
68628     36 months
98758     36 months
13070     36 months
45836     36 months
52330     36 months
48243     36 months


In [48]:
loansData['Loan.Length'].value_counts()

36 months    1952
60 months     548
Name: Loan.Length, dtype: int64

### Conclusion of Step 2
* Here we used the techniques we learned in Step 1.
 * Pull out a column from a data frame
 * Operate on it, perform some transformations
 * Replace the column in the original dataframe with this new column  
* We applied them to removing the ' months' suffix in the Loan.Length column.

## Step 3: Remove bad data

In [49]:
loansData['Monthly.Income'].describe()

count      2499.000000
mean       5688.931321
std        3963.118185
min         588.500000
25%        3500.000000
50%        5000.000000
75%        6800.000000
max      102750.000000
Name: Monthly.Income, dtype: float64

First remove implausible values.  We see the max value to be 102750.  
This is a MONTHLY income of 100K dollars, which is certainly possible, but ....  
highly implausible for a person seeking a loan of a few 10's of K dollars, i.e. implausible in this context.  

In [50]:
loansData['Monthly.Income'].idxmax() # find the place where the max occurs

54487

But there's a better way - a row filter i.e. an expression used as a way to restrict the rows in a dataframe.  
In our case we want to eliminate rows above 100K dollars. i.e. only keep those less than 100K dollars.

In [51]:
loansData['Monthly.Income'][loansData['Monthly.Income'] < 100000]

81174      6541.67
99592      4583.33
80059     11500.00
15825      3833.33
33182      3195.00
62403      4891.67
48808      2916.67
22090     13863.42
76404      3150.00
15867      5000.00
94971      3575.00
36911      4250.00
41200     14166.67
83869      9166.67
53853     11250.00
21399      5416.67
62127      9000.00
23446      4333.33
44987      2733.33
17977      3750.00
86099      6666.67
99483      6250.00
28798      4100.00
24168      6833.33
10356      4500.00
46027      5416.67
2238       2964.17
65278      2916.67
4227       4167.00
50182      3958.33
            ...   
84265      6750.00
80231      5166.67
49533      5500.00
102514     3833.00
78618      8333.33
86953      4583.33
80129      8583.33
85216     25000.00
38247      5416.67
91245      4833.33
53041      5000.00
63051      9250.00
14446      7075.50
68628      3416.67
98758      6666.67
13070      7083.33
45836      2464.37
52330      8000.00
48243      3833.33
63256      5166.67
42124      4500.00
78043      3

In [52]:
loansData['Monthly.Income'].describe()

count      2499.000000
mean       5688.931321
std        3963.118185
min         588.500000
25%        3500.000000
50%        5000.000000
75%        6800.000000
max      102750.000000
Name: Monthly.Income, dtype: float64

In [53]:
ldlt100 = ldb[ldb['Monthly.Income'] < 100000]

In [54]:
ldlt100

index  Amount.Requested  Amount.Funded.By.Investors  Interest.Rate  \
0      81174             20000                    20000.00           8.90   
1      99592             19200                    19200.00          12.12   
2      80059             35000                    35000.00          21.98   
3      15825             10000                     9975.00           9.99   
4      33182             12000                    12000.00          11.71   
5      62403              6000                     6000.00          15.31   
6      48808             10000                    10000.00           7.90   
7      22090             33500                    33450.00          17.14   
8      76404             14675                    14675.00          14.33   
9      15867              7000                     7000.00           6.91   
10     94971              2000                     2000.00          19.72   
11     36911             10625                    10625.00          14.27   
12     41200             28000                    27975.00          21.67   
13     83869             35000                    34950.00           8.90   
14     53853              9600                     9600.00           7.62   
15     21399             25000                    24975.00          15.65   
16     62127             10000                    10000.00          12.12   
17     23446             14000                    13900.25          10.37   
18     44987             10000                    10000.00           9.76   
19     17977              5200                     5175.00           9.99   
20     86099             22000                    21975.00          21.98   
21     99483             30000                    30000.00          19.05   
22     28798              6500                     6500.00          17.99   
23     24168             17400                    17400.00          11.99   
24     10356              4000                     4000.00          16.82   
25     46027              7200                     7200.00           7.90   
26      2238              8000                     8000.00          14.42   
27     65278              8000                     8000.00          15.31   
28      4227              3000                     3000.00           8.59   
29     50182             14500                    14500.00           7.90   
...      ...               ...                         ...            ...   
2470   84265             20000                    20000.00          22.95   
2471   80231             19000                    19000.00           7.90   
2472   49533             17300                    17250.00          22.45   
2473  102514              7000                      711.54          15.13   
2474   78618              7200                     7200.00          18.75   
2475   86953             10000                    10000.00          14.09   
2476   80129              4000                     3925.00          14.09   
2477   85216             17500                    17500.00           8.90   
2478   38247             20000                    20000.00          11.71   
2479   91245             16200                    16200.00          15.80   
2480   53041             10000                    10000.00           6.03   
2481   63051             27000                    27000.00           6.62   
2482   14446              4500                     4475.00           7.51   
2483   68628             15875                    15875.00          14.33   
2484   98758             15000                    15000.00          10.16   
2485   13070             25000                    24950.00          10.75   
2486   45836              7000                     7000.00          17.27   
2487   52330             15000                    15000.00          19.99   
2488   48243             17000                    17000.00          15.81   
2489   63256             19075                    19075.00          18.75   
249

In [55]:
len(ldlt100)

2498

Now drop any rows that have 'NA' values ie data not available.  
In database terminiology these would be 'NULL' values.

In [56]:
ldb2 = ldlt100.dropna()

In [57]:
ldb2

index  Amount.Requested  Amount.Funded.By.Investors  Interest.Rate  \
0      81174             20000                    20000.00           8.90   
1      99592             19200                    19200.00          12.12   
2      80059             35000                    35000.00          21.98   
3      15825             10000                     9975.00           9.99   
4      33182             12000                    12000.00          11.71   
5      62403              6000                     6000.00          15.31   
6      48808             10000                    10000.00           7.90   
7      22090             33500                    33450.00          17.14   
8      76404             14675                    14675.00          14.33   
9      15867              7000                     7000.00           6.91   
10     94971              2000                     2000.00          19.72   
11     36911             10625                    10625.00          14.27   
12     41200             28000                    27975.00          21.67   
13     83869             35000                    34950.00           8.90   
14     53853              9600                     9600.00           7.62   
15     21399             25000                    24975.00          15.65   
16     62127             10000                    10000.00          12.12   
17     23446             14000                    13900.25          10.37   
18     44987             10000                    10000.00           9.76   
19     17977              5200                     5175.00           9.99   
20     86099             22000                    21975.00          21.98   
21     99483             30000                    30000.00          19.05   
22     28798              6500                     6500.00          17.99   
23     24168             17400                    17400.00          11.99   
24     10356              4000                     4000.00          16.82   
25     46027              7200                     7200.00           7.90   
26      2238              8000                     8000.00          14.42   
27     65278              8000                     8000.00          15.31   
28      4227              3000                     3000.00           8.59   
29     50182             14500                    14500.00           7.90   
...      ...               ...                         ...            ...   
2470   84265             20000                    20000.00          22.95   
2471   80231             19000                    19000.00           7.90   
2472   49533             17300                    17250.00          22.45   
2473  102514              7000                      711.54          15.13   
2474   78618              7200                     7200.00          18.75   
2475   86953             10000                    10000.00          14.09   
2476   80129              4000                     3925.00          14.09   
2477   85216             17500                    17500.00           8.90   
2478   38247             20000                    20000.00          11.71   
2479   91245             16200                    16200.00          15.80   
2480   53041             10000                    10000.00           6.03   
2481   63051             27000                    27000.00           6.62   
2482   14446              4500                     4475.00           7.51   
2483   68628             15875                    15875.00          14.33   
2484   98758             15000                    15000.00          10.16   
2485   13070             25000                    24950.00          10.75   
2486   45836              7000                     7000.00          17.27   
2487   52330             15000                    15000.00          19.99   
2488   48243             17000                    17000.00          15.81   
2489   63256             19075                    19075.00          18.75   
249

In [58]:
len(ldb2)

2497

So we dropped one row that had an NA value somewhere.

### Conclusion of step 3
* We used techniques we learned in step 1 to pick out a columns and operate on it
* We also learned how to filter the data based on expressions involving column values
* Finally we learnt how to drop NA values

Now that we have removed the "bad" data, let's take on the final data-cleaning task for this data set - converting the ranges 
to single integers.

## Step 4: Change FICO range to a single value

Note that the FICO values are given in a range which is in the form of a string that looks like lowerlimit-upperlimit, eg 720-724.  
We want to convert these values to a single int value representing the lower limit. e.g. 720 in the above example.


In [59]:
ficostr = ldb2['FICO.Range'] 

In [60]:
ficostr[0:10]

0    735-739
1    715-719
2    690-694
3    695-699
4    695-699
5    670-674
6    720-724
7    705-709
8    685-689
9    715-719
Name: FICO.Range, dtype: object

In [61]:
ficostr[0]

'735-739'

In [62]:
ficoint = [ int(x.split('-')[0]) for x in ficostr ]

In [63]:
ficoint[0:10]

[735, 715, 690, 695, 695, 670, 720, 705, 685, 715]

In [64]:
ldb2['FICO.Range'] = ficoint

/home/marco/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [65]:
len(ficoint)

2497

In [66]:
ldb2['FICO.Range']

0       735
1       715
2       690
3       695
4       695
5       670
6       720
7       705
8       685
9       715
10      670
11      665
12      670
13      735
14      725
15      730
16      695
17      740
18      730
19      760
20      665
21      695
22      665
23      695
24      670
25      705
26      675
27      675
28      765
29      760
       ... 
2470    665
2471    770
2472    685
2473    650
2474    660
2475    675
2476    675
2477    730
2478    725
2479    680
2480    760
2481    810
2482    720
2483    675
2484    690
2485    765
2486    665
2487    660
2488    685
2489    670
2490    710
2491    720
2492    710
2493    675
2494    685
2495    705
2496    740
2497    680
2498    675
2499    670
Name: FICO.Range, Length: 2497, dtype: int64

#### Conclusion of Step 4

* We used techniques similar to the ones in Step 1
 * we picked a column - the FICO.Range column
 * we split the values on the separator '-'
 * we picked the fist value, i.e. the lower limit
 * we converted it to an int

Now we have a dataset that we can use for our data exploration and analysis